larray/core/session.py

```diff
@@ -1393,87 +1394,85 @@ def display(k, v, is_metadata=False):
-    For example, a user don't need to remember that a new label has been added to
-    an axis of a given array somewhere earlier in the code (in a part of the model
-    written by a colleague). Therefore, these restrictions reduces the risk to deal
-    with unexpected error messages (like 'Incompatible Axes') and make it easier to
-    work in team.
-
-    Parameters
-    ----------
-    filepath: str, optional
-        Path where items have been saved. This can be either the path to a single file, a path to
-        a directory containing .csv files or a pattern representing several .csv files.
-    meta : list of pairs or dict or OrderedDict or Metadata, optional
-        Metadata (title, description, author, creation_date, ...) associated with the array.
-        Keys must be strings. Values must be of type string, int, float, date, time or datetime.
-
-    Notes
-    -----
-    The 'autocomplete' is a feature in which a software predicts the rest of a variable or function
-    name after a user typed the first letters. This feature allows users to use longer but meaningful
-    variable or function names (like 'population_be' instead of 'pbe') and to avoid creating an unwanted
-    new variable by misspelling the name of a given variable (e.g. typing 'poplation = something'
-    ('population' without u) will create a new variable instead of modifying the 'population' variable).
 
+# XXX: the name of the class below is not really important since it will serve as base
+#      for the LazySession and users are expected to inherit from the LazySession when
+#      they define their own classes
+class TypedSession(Session):
+    """
     Examples
     --------
-    >>> class ModelVariables(FrozenSession):
-    ...     LAST_AGE = 120
-    ...     FIRST_OBS_YEAR = 1991
-    ...     LAST_PROJ_YEAR = 2070
-    ...     AGE = Axis('age=0..{}'.format(LAST_AGE))
-    ...     GENDER = Axis('gender=male,female')
-    ...     TIME = Axis('time={}..{}'.format(FIRST_OBS_YEAR, LAST_PROJ_YEAR))
-    ...     CHILDREN = AGE[0:17]
-    ...     ELDERS = AGE[65:]
-    ...     population = zeros((AGE, GENDER, TIME))
-    ...     births = zeros((AGE, GENDER, TIME))
-    ...     deaths = zeros((AGE, GENDER, TIME))
-    >>> m = ModelVariables()
-    >>> m.names    # doctest: +NORMALIZE_WHITESPACE
-    ['AGE', 'CHILDREN', 'ELDERS', 'FIRST_OBS_YEAR', 'GENDER', 'LAST_AGE', 'LAST_PROJ_YEAR', 'TIME', 'births',
-    'deaths', 'population']
+    Content of file 'model_variables.py'
+
+    >>> # ==== MODEL VARIABLES ====
+    >>> class ModelVariables(TypedSession):
+    ...     FIRST_OBS_YEAR = int
+    ...     FIRST_PROJ_YEAR = int
+    ...     LAST_PROJ_YEAR = int
+    ...     AGE = Axis
+    ...     GENDER = Axis
+    ...     TIME = Axis
+    ...     G_CHILDREN = Group
+    ...     G_ADULTS = Group
+    ...     G_OBS_YEARS = Group
+    ...     G_PROJ_YEARS = Group
+    ...     population = ArrayDef(('AGE', 'GENDER', 'TIME'))
```

Comment:
I have some vague aversion about mixing arrays and parameters in the same Session. I cannot word it yet. Maybe it is just because that's not how you work in "demo", but I think there is something else behind my gut feeling. Technically I don't think we need anything different but I wonder what is better to show as an example: something simple but not optimal, or the "better" (but more complex) organisation.

Reply:
What do you consider as parameters? Everything that doesn't change at runtime or read-only scalars?

Reply:
Hehe, this innocent-looking question is actually a very good question, because it would help to have a clear idea and precise definitions of this both for our examples/documentation and our users models. I think I would define a constant as something which never ever changes (across all variants of a model) and a parameter as anything which can change from one variant or simulation to the next but is constant over the simulation. This might be a very personal take on the issue, I don't know. Unsure where to draw the line between input data and parameter though.